src/Containers/Microsoft.NET.Build.Containers/Registry.cs

```diff
@@ -339,68 +343,186 @@ public async Task<string> DownloadBlobAsync(string repository, Descriptor descri
         return localPath;
     }
 
-    public async Task PushAsync(Layer layer, string repository, Action<string> logProgressMessage, CancellationToken cancellationToken)
+    public async Task PushAsync(Layer layer, string repository, Action<LogMessage> logProgressMessage, CancellationToken cancellationToken)
     {
         cancellationToken.ThrowIfCancellationRequested();
         string digest = layer.Descriptor.Digest;
 
         using (FileStream contents = File.OpenRead(layer.BackingFile))
         {
-            await UploadBlobAsync(repository, digest, contents, cancellationToken).ConfigureAwait(false);
+            await UploadBlobAsync(repository, digest, contents, logProgressMessage, cancellationToken).ConfigureAwait(false);
         }
     }
 
-    private async Task<UriBuilder> UploadBlobChunkedAsync(string repository, string digest, Stream contents, HttpClient client, UriBuilder uploadUri, CancellationToken cancellationToken)
+    /// <summary>
+    /// Captures the data needed to finalize an upload
+    /// </summary>
+    private record UploadFinalizeInformation(UriBuilder uploadUri, string digest);
+
+    /// <summary>
+    /// Upload a blob to the registry using chunks. The chunks are uploaded sequentially, and the size of the chunks is determined by the UploadInformation derived from the 'Start' call
+    /// </summary>
+    private async Task<UploadFinalizeInformation> UploadBlobChunkedAsync(string repository, Stream contents, HttpClient client, UploadInformation uploadInfo, Action<LogMessage> logProgressMessage, CancellationToken cancellationToken)
     {
         cancellationToken.ThrowIfCancellationRequested();
-        Uri patchUri = uploadUri.Uri;
-        var localUploadUri = new UriBuilder(uploadUri.Uri);
-        localUploadUri.Query += $"&digest={Uri.EscapeDataString(digest)}";
+        Uri patchUri = uploadInfo.uploadUri.Uri;
+        contents.Seek(0, SeekOrigin.Begin);
 
-        // TODO: this chunking is super tiny and probably not necessary; what does the docker client do
-        //       and can we be smarter?
-
-        byte[] chunkBackingStore = new byte[MaxChunkSizeBytes];
+        using IncrementalHash hash = IncrementalHash.CreateHash(HashAlgorithmName.SHA256);
```

Comment:
I added incremental hash generation for chunks because at one point I thought hash calculation was getting modified. Trace logging proved this wasn't so - this can probably safely be removed at this point.

Reply:
Will it be removed then? or moved to DEBUG at least?